### PR TITLE
Allocator: Expand stack space when stealing virtual address space

### PR DESCRIPTION
**Pull request by Sonicadvance1**
If we take all of the stack space then the auto expanding stack doesn't work and we get stuck with a small stack that breaks thunks.

### CODE REVIEW

File: External/FEXCore/Source/Utils/Allocator.cpp [hunk 1]
```diff
@@ -145,6 +145,8 @@ namespace FEXCore::Allocator {
   #define STEAL_LOG(...) // fprintf(stderr, __VA_ARGS__)
 
   std::vector<MemoryRegion> StealMemoryRegion(uintptr_t Begin, uintptr_t End) {
+    void *StackLocation = alloca(0);
+    uintptr_t StackLocation_u64 = reinterpret_cast<uintptr_t>(StackLocation);
```
Comment (neobrain) [inline External/FEXCore/Source/Utils/Allocator.cpp:149]:
```suggestion
    void * const StackLocation = alloca(0);
    const uintptr_t StackLocation_u64 = reinterpret_cast<uintptr_t>(StackLocation);
```

File: External/FEXCore/Source/Utils/Allocator.cpp [hunk 2]
```diff
@@ -145,6 +145,8 @@ namespace FEXCore::Allocator {
   #define STEAL_LOG(...) // fprintf(stderr, __VA_ARGS__)
 
   std::vector<MemoryRegion> StealMemoryRegion(uintptr_t Begin, uintptr_t End) {
+    void *StackLocation = alloca(0);
```
Comment (neobrain) [inline External/FEXCore/Source/Utils/Allocator.cpp:148]: Is it safe to call this at the top of the function, rather than after all other variables are declared? If any of the variables are spilled, the stack location would change by the time we use `StackLocation`.

Comment (neobrain) [inline External/FEXCore/Source/Utils/Allocator.cpp:148]: Does this query the position of the last byte already allocated on stack or the position of the *next* byte that would be allocated for a non-zero size argument? Comment might be useful here. Not sure we can safely rely on the behavior of `alloca(0)` to begin with though, so just using `alloca(1)` or taking the reference to a local variable might be better.

Comment (Sonicadvance1) [reply External/FEXCore/Source/Utils/Allocator.cpp]: This is safe since we are only using the stack location to find which VMA is holding our stack when parsing through `/proc/self/maps`.

Once we find the VMA that contains the stack range we know that we have the right one and allocate the prior one so our stack can keep growing.

The way this code works is a bit hard to parse since I didn't write it. I'll likely rewrite this in the future to make it less of a nightmare.

Comment (Sonicadvance1) [reply External/FEXCore/Source/Utils/Allocator.cpp]: `alloca(0)` returns the current stack pointer, so the last byte already allocated on the stack.
We use this pattern in a few other places of code since we need to pull the stack location for various reasons.
By spec this function will return an address even if it results in a stack overflow, since it doesn't do any checking of what is passed in

File: External/FEXCore/Source/Utils/Allocator.cpp [hunk 3]
```diff
@@ -240,6 +248,22 @@ namespace FEXCore::Allocator {
           STEAL_LOG("[%d] ParseEnd; RegionBegin: %016lX RegionEnd: %016lX\n", __LINE__, RegionBegin, RegionEnd);
 
           State = ScanEnd;
+
+          // If the previous map's ending and the region we just parsed overlaps the stack then we need to save the stackmapping.
```
Comment (neobrain) [inline External/FEXCore/Source/Utils/Allocator.cpp:252]:
```suggestion
          // If the previous map's ending and the region we just parsed overlap the stack then we need to save the stack mapping.
```

File: External/FEXCore/Source/Utils/Allocator.cpp [hunk 4]
```diff
@@ -240,6 +248,22 @@ namespace FEXCore::Allocator {
...
+          // Otherwise we will only have ~144KB stack size which crashes quickly.
```
Comment (neobrain) [inline External/FEXCore/Source/Utils/Allocator.cpp:253]: Where does the 144kB value actually come from?

File: External/FEXCore/Source/Utils/Allocator.cpp [hunk 5]
```diff
@@ -211,7 +215,7 @@ namespace FEXCore::Allocator {
 
           auto MapBegin = std::max(RegionEnd, Begin);
           auto MapEnd = std::min(RegionBegin, End);
-          
+
```
Comment (neobrain) [inline External/FEXCore/Source/Utils/Allocator.cpp:218]: This seems better placed here:
```suggestion
          // Store the previous map location
          PreviousMapEnd = MapEnd;
```

Alternatively, MapEnd should be made `const` to make it easier to see you're just storing the unmodified value later on.

File: External/FEXCore/Source/Utils/Allocator.cpp (hunk 2 above)
Comment (neobrain) [reply External/FEXCore/Source/Utils/Allocator.cpp]: > alloca(0) returns the current stack pointer, so the last byte already allocated on the stack.

As a followup, is it the last *byte* or the last *value*?

>  By spec this function will return an address even if it results in a stack overflow, since it doesn't do any checking of what is passed in

Is the behavior guaranteed for the argument 0 though? `malloc(0)` for instance will return nullptr on some implementations, which would silently break here if a future libc update adopted similar behavior for alloca.

File: External/FEXCore/Source/Utils/Allocator.cpp [hunk 6]
```diff
@@ -240,6 +248,22 @@ namespace FEXCore::Allocator {
           STEAL_LOG("[%d] ParseEnd; RegionBegin: %016lX RegionEnd: %016lX\n", __LINE__, RegionBegin, RegionEnd);
 
           State = ScanEnd;
+
+          // If the previous map's ending and the region we just parsed overlaps the stack then we need to save the stackmapping.
+          // Otherwise we will only have ~144KB stack size which crashes quickly.
+          if (PreviousMapEnd <= StackLocation_u64 && RegionEnd > StackLocation_u64) {
```
Comment (neobrain) [inline External/FEXCore/Source/Utils/Allocator.cpp:254]: How did you come up with this check? Specifically, why does `PreviousMapEnd <= StackLocation_u64` check that the previous map overlaps with the stack? (My understanding being that the stack ends at `StackLocation_u64` and begins at an address higher than that)

File: External/FEXCore/Source/Utils/Allocator.cpp (hunk 4 above)
Comment (Sonicadvance1) [reply External/FEXCore/Source/Utils/Allocator.cpp:253]: From me inspecting a crash from thunks where that was the amount of stack space used up to that point of crash.
Can change the comment to some non-constant value like "We will only have an itty-bitty stack size..."

File: External/FEXCore/Source/Utils/Allocator.cpp (hunk 2 above)
Comment (Sonicadvance1) [reply External/FEXCore/Source/Utils/Allocator.cpp]: From the spec wording. probably.
If alloca starts returning zero on us then all of our guest threads immediately break/crash which will be a stronger indication that something is wrong than us failing to allocate more stack space. Since they do the same thing.

File: External/FEXCore/Source/Utils/Allocator.cpp (hunk 5 above)
Comment (Sonicadvance1) [reply External/FEXCore/Source/Utils/Allocator.cpp:218]: Moved and made const.

File: External/FEXCore/Source/Utils/Allocator.cpp (hunk 4 above)
Comment (Sonicadvance1) [reply External/FEXCore/Source/Utils/Allocator.cpp]: Changed comment to not explicitly talk about a stack size, which will be variable.

File: External/FEXCore/Source/Utils/Allocator.cpp (hunk 3 above)
Comment (Sonicadvance1) [reply External/FEXCore/Source/Utils/Allocator.cpp]: Updated.

File: External/FEXCore/Source/Utils/Allocator.cpp (hunk 6 above)
Comment (Sonicadvance1) [reply External/FEXCore/Source/Utils/Allocator.cpp:254]: This is absolutely bonkers how the logic in this function works.
First step is that it parses the region below the stack space, mmap'ing it with no privileges. Saving the `PreviousMapEnd` variable to the start of the mapping address.

It then parses the stack mapping in the `State == ParseEnd` step.
This is where the `if (PreviousMapEnd <= StackLocation_u64 && RegionEnd > StackLocation_u64)` comparison comes in.
First check ensuring that the previous mapping end address is less than the current stack pointer. Ensuring we haven't progressed beyond the stack comparison point.

Second check ensuring that the /current/ parsed region is higher than the current stack pointer, ensuring that this full region is the stack region.
As long as everything is allocated in order like expected then this holds true that it is the stack.

As an additional step of safety, when we pull the last allocation off the vector of allocations, we double check that `Ptr+Size` is the start of the mapping, ensuring the regions are contiguous with each other.

Absolute nightmare, but we'll rewrite this code to be easier to parse.

**Comment by Sonicadvance1** [comment]
This one still needs a final review since otherwise thunks don't work in most Proton titles.

**Review by lioncash — APPROVED** [review]
